contracts/told_ya/src/lib.cairo

```diff
@@ -1,69 +1,252 @@
+        events_id: Array<felt252>,
+        predictions: LegacyMap::<felt252, Prediction>,
+        predictions_id: Array<felt252>,
+        user_predictions_id: LegacyMap::<ContractAddress, Array<felt252>>,
         #[substorage(v0)]
-        ownable: OwnableComponent::Storage,
-        balance: felt252
+        ownable: super::OwnableComponent::Storage
     }
 
     #[event]
     #[derive(Drop, starknet::Event)]
     enum Event {
-        OwnableEvent: OwnableComponent::Event
+        OwnableEvent: super::OwnableComponent::Event
     }
 
     #[constructor]
-    fn constructor(ref self: ContractState, initial_owner: ContractAddress){
-        self.ownable.initializer(initial_owner);
+    fn constructor(ref self: ContractState){
+        let caller_address = starknet::get_caller_address();
+        self.ownable.initializer(caller_address);
     }
 
     #[abi(embed_v0)]
-    impl HelloStarknetImpl of super::IHelloStarknet<ContractState> {
-        fn increase_balance(ref self: ContractState, amount: felt252) {
-            assert(amount != 0, 'Amount cannot be 0');
-            self.balance.write(self.balance.read() + amount);
+    impl ToldYaImpl of super::IToldYa<ContractState> {
+
+        fn create_event(ref self: ContractState, name: felt252, predictions_deadline: felt252, event_datetime: felt252, type_: felt252) -> Event_ {
+            // 1) Verifying caller is owner
+            self.ownable.assert_only_owner();
+
+            // 2) Hashing the new event
+            let hash_state = PoseidonTrait::new();
+            let hash_result: felt252 = hash_state.update(name).update(predictions_deadline).update(event_datetime).update(type_).finalize();            
+
+            // 3) Creating the instance of the new event
+            let new_event: Event_ = Event_ {
+                identifier: hash_result, 
+                name: name,
+                predictions_deadline: predictions_deadline,
+                event_datetime: event_datetime,
+                type_: type_
+            };
+
+            // 4) Storing the new event in Storage
+            self.events.write(new_event.identifier, new_event);
+
+            // 5) Adding the id to events_id
+            let mut events_id: Array<felt252> = self.events_id.read();
+            events_id.append(new_event.identifier);
+            self.events_id.write(events_id);
+
+            //6) Returning the new event
+            new_event
+        }
+
+        fn create_prediction(ref self: ContractState, event_identifier: felt252, value: felt252) -> Prediction {
+
+            // 1) Checking if event_identifier is valid
+            let mut events_id: Array<felt252> = self.events_id.read();
+            let mut event_id_is_valid: bool = false;
+            while !events_id.is_empty(){
+                let event_id = events_id.pop_front().unwrap();
+                if event_identifier == event_id {
+                    event_id_is_valid = true;
+                }
```

Comment:
Yes I agree, it would be much better to do it that way.

I did it like this because I don't know how the self.storage.read(key) behave if the key is not in the LegacyMap list of keys. I haven't found it in the documentation. But I am sure reading the source code would allow us to understand the behaviour.

Reply:
I see, don't worry we'll leave it this way for now, it will be part of a future refactor or smth